### ISSUE TITLE
Meaning of supported application protocols

### ISSUE THREAD

**Issue by jsrojas** [ntop/nDPI#841]
Hello everyone

I have one quick question regarding the meaning of supported application protocols. When nDPI inspects a packet and classifies it as Apple what does it mean? is it a communication with an Apple application from a mac device?

The same happens when the packet is classified as Google, is it a search query on Google search engine or is it something else?

I ask this because I see that nDPI can differentiate between several apps like Apple.Itunes, Apple.Icloud, GoogleDocs, GoogleDrive,etc. But that is why I am wondering what does it mean when the packet is classified as Apple or Google only (without any further specification).

Thanks in advance.

**Comment by aouinizied** [comment]
nDPI is a deep packet inspection library. Consequently, it combines several classical techniques (e.g: IP matching, metadata content matching, port guessing) and more advanced techniques for specific dissectors.

Now let's take web (HTTP, HTTPS) use case. 
When fine-grained service identification is possible (GoogleDocs, GoogleDrive) such in the following scenario (example):

- TCP session established and the fouth packet hit nDPI with some metadata in it (TLS Server name Identifier, HTTP UserAgent, etc).
- nDPI in this case performs content matching and can identify finely which service is in it.
Here an example of patterns used for GoogleDrive:
https://github.com/ntop/nDPI/blob/97fb50e837da46bac7bedeeb170cace6dd3a912c/src/lib/ndpi_content_match.c.inc#L8767

Now in another scenario, when there is only establishement and for some reason we didn't observe a packet with metadata.
Another possible scenario is that we have metadata but we don't know to which service it belongs.
The fallback in this cases is to use IP matching (ServerIPAddress) which will indicate it's a Google/Apple but not the exact service contacted.
Here an example of the IP adresses Pool used for Google.
https://github.com/ntop/nDPI/blob/97fb50e837da46bac7bedeeb170cace6dd3a912c/src/lib/ndpi_content_match.c.inc#L1050

To sum up, let's say that it is an incremental process that goes deeper when possible.
So at first packet level, nDPI can say that it's Google, if there is dissectable metadata data observed on the rest of the flow' packets, nDPI will affine its classification according to it. If not, it's always better to say it's Google than Unknown.

Finally, this is a coarse understanding of how things works,  Luca or one member of ntop team may correct my answer if I'm missing something. :)

Best regards,
Zied

**Comment by aouinizied** [comment]
Very useful slides provided by ntop team that helps to understand how encrypted traffic is handled and classified can be found here: https://www.ntop.org/wp-content/uploads/2020/01/EncryptedTrafficAnalysis.pdf

**Comment by jsrojas** [comment]
Hello Zied

As always thank you so much for your accurate, complete and quick responses (was not expecting you to come to my rescue again). Let's wait for any additional information that the ntop team members can offer.

Also, thank you for the slides, honestly it has been quite difficult to find detailed documentation about nDPI code or its functionality if you want to go deeper into it, so it really helps a lot :D :D 

Best regards
Juan

**Comment by emarshswe** [comment]
This is a question not an issue and should be tagged as such.
